src/walk.rs

```diff
@@ -160,6 +162,161 @@ pub fn scan(path_vec: &[PathBuf], pattern: Arc<Regex>, config: Arc<Config>) -> R
     }
 }
 
+/// Wrapper for the receiver thread's buffering behavior.
+struct ReceiverBuffer<W> {
+    /// The configuration.
+    config: Arc<Config>,
+    /// The ^C notifier.
+    wants_to_quit: Arc<AtomicBool>,
+    /// Receiver for worker results.
+    rx: Receiver<WorkerResult>,
+    /// Standard output.
+    stdout: W,
+    /// The current buffer mode.
+    mode: ReceiverMode,
+    /// The deadline to switch to streaming mode.
+    deadline: Instant,
+    /// The buffer of quickly received paths.
+    buffer: Vec<PathBuf>,
+    /// Result count.
+    num_results: usize,
+}
+
+impl<W: Write> ReceiverBuffer<W> {
+    /// Create a new receiver buffer.
+    fn new(
+        config: Arc<Config>,
+        wants_to_quit: Arc<AtomicBool>,
+        rx: Receiver<WorkerResult>,
+        stdout: W,
+    ) -> Self {
+        let max_buffer_time = config.max_buffer_time.unwrap_or(DEFAULT_MAX_BUFFER_TIME);
+        let deadline = Instant::now() + max_buffer_time;
+
+        Self {
+            config,
+            wants_to_quit,
+            rx,
+            stdout,
+            mode: ReceiverMode::Buffering,
+            deadline,
+            buffer: Vec::with_capacity(MAX_BUFFER_LENGTH),
+            num_results: 0,
+        }
+    }
+
+    /// Process results until finished.
+    fn process(&mut self) -> ExitCode {
+        loop {
+            if let Err(ec) = self.poll() {
+                return ec;
+            }
+        }
+    }
+
+    /// Receive the next worker result.
+    fn recv(&self) -> Result<WorkerResult, RecvTimeoutError> {
+        match self.mode {
+            ReceiverMode::Buffering => {
+                // Wait at most until we should switch to streaming
+                let now = Instant::now();
+                self.deadline
+                    .checked_duration_since(now)
+                    .ok_or(RecvTimeoutError::Timeout)
+                    .and_then(|t| self.rx.recv_timeout(t))
```

Comment:
If/once [`recv_deadline`](https://doc.rust-lang.org/stable/std/sync/mpsc/struct.Receiver.html#method.recv_deadline) (Tracked by https://github.com/rust-lang/rust/issues/46316) is stabilized, that would probably be a better fit here.

Reply:
Yeah for sure.  (Do you know what the process for requesting stabilization is?)

Funnily enough `recv_timeout()` is [implemented in terms of `recv_deadline()`](https://github.com/rust-lang/rust/blob/68b554e6af18726fe6fa8de2134c59c441e0b019/library/std/src/sync/mpsc/mod.rs#L1276) so here I'm emulating `recv_deadline()` with `recv_timeout()` which is emulated by `recv_deadline()`.

Reply:
I don't know exactly what the process for that is. From the tracking issue it looks like the biggest open question is the name of the function.

Reply:
Huh well here's the news on that: https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Stabilizing.20.60recv_deadline.28.29.60

Kinda amazed that https://github.com/rust-lang/rust/issues/39364 is still open.  I guess we should switch to `crossbeam-channels` before merging this PR.

Reply:
Switched to `crossbeam-channel`